### ISSUE TITLE
Trying to build yolo_cpp_dll_no_gpu

### ISSUE THREAD

**Issue by Nfear** [AlexeyAB/darknet#162]
It says `Cannot open include file: 'cuda_runtime.h': No such file or directory`.

I have CUDA installed (GPU doesn't support CUDA) and the file is located at `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\include`.

EDIT:

I had the Property Pages Configuration set to release, not the Build Configuration. Changing both to release fixed it.

**Comment by VanitarNordic** [comment]
You will get old on training deep learning models such as YOLO on a CPU. forget it in all costs. One training can take up to few months on a CPU!

**Comment by AlexeyAB** [comment]
I fixed it. Now `yolo_cpp_dll_no_gpu.sln` can be built using any: Release or Debug.

**Comment by changss** [comment]
@AlexeyAB
 void test_detector_cpu(char *datacfg, char *cfgfile, char *weightfile, char *filename, float thresh);

**Comment by shartoo** [comment]
I'm trying to compile `yolo_cpp_dll_no_gpu.sln` on a windows xp with vs2010 and without cuda but failed with error below
![微信图片_20191106171746](https://user-images.githubusercontent.com/5716282/68285054-86e00700-00b9-11ea-893f-3ada823e87c9.png)

the error messgae is `darknet-master/include/darknet.h(39): fatal error C1083: Could not open include file :“cuda_runtime.h”: No such file or directory`

**Comment by AlexeyAB** [comment]
@shartoo 

1. Open solution yolo_cpp_dll.sln in MSVS
2. Open file darknet.h
3. Find `#ifdef GPU` in line 37
4. press right mouse-button on `GPU`
5. press `find definition` / `go to definition`
6. show me screenshot of result

**Comment by shartoo** [comment]
`Could not find definition of GPU`

![微信截图_20191107103529](https://user-images.githubusercontent.com/5716282/68355467-63b16800-014a-11ea-986c-b258095f3d10.png)

**Comment by shartoo** [comment]
Here is the full output error

```
Here is the full output error

```
------ regenerated: project: yolo_cpp_dll_no_gpu, Configuration : Release Win32 ------
 2019-11-7 10:39:44。
The project contains configuration of ToolsVersion="14.0" ，but current version of  MSBuild unsupport 。The project will regard  as configuration of ToolsVersion="4.0" 
_PrepareForClean:
  cleaning “Release\yolo_cpp_dll_no_gpu.lastbuildstate”。
InitializeBuildStatus:
  “Release\yolo_cpp_dll_no_gpu.unsuccessfulbuild”executing  Touch task。
ClCompile:
  activations.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open:“cuda_runtime.h”: No such file or directory
  activation_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  art.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  avgpool_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  batchnorm_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  blas.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  box.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  captcha.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  cifar.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  classifier.c

  col2im.c
..\..\src\col2im.c(47): error C2143: grammer error : lack of “;”(before“type”)
..\..\src\col2im.c(48): error C2065: “i”: Undeclared definition
..\..\src\col2im.c(48): error C2065: “i”: Undeclared definition
..\..\src\col2im.c(48): error C2065: “i”: Undeclared definition
..\..\src\col2im.c(49): error C2065: “i”: Undeclared definition
..\..\src\col2im.c(53): error C2054: insert "(" after “inline”
..\..\src\col2im.c(53): error C2085: “is_a_ge_zero_and_a_lt_b”: not in the formal parameters list
..\..\src\col2im.c(53): error C2143: grammer error : lack of“;”(before“{”)
..\..\src\col2im.c(66): error C2143: grammer error : lack of“;”(before “const”)
..\..\src\col2im.c(68): error C2143: grammer error : lack of“;”(before “const”)
..\..\src\col2im.c(70): error C2143: grammer error : lack of“;”(before “const”)
..\..\src\col2im.c(71): error C2143: grammer error : lack of“;”(before “type”)
..\..\src\col2im.c(72): error C2065: “channel”: Undeclared definition
..\..\src\col2im.c(72): error C2065: “channel”: Undeclared definition
..\..\src\col2im.c(72): error C2065: “channel_size”: Undeclared definition
..\..\src\col2im.c(73): error C2065: “kernel_row”: Undeclared definition
..\..\src\col2im.c(73): error C2065: “kernel_row”: Undeclared definition
..\..\src\col2im.c(73): error C2065: “kernel_row”: Undeclared definition
..\..\src\col2im.c(74): error C2065: “kernel_col”: Undeclared definition
..\..\src\col2im.c(74): error C2065: “kernel_col”: Undeclared definition
..\..\src\col2im.c(74): error C2065: “kernel_col”: Undeclared definition
..\..\src\col2im.c(75): error C2065: “kernel_row”: Undeclared definition
..\..\src\col2im.c(76): error C2065: “output_rows”: Undeclared definition
..\..\src\col2im.c(76): error C2065: “output_h”: Undeclared definition
..\..\src\col2im.c(76): error C2065: “output_rows”: Undeclared definition
..\..\src\col2im.c(76): error C2065: “output_rows”: Undeclared definition
..\..\src\col2im.c(77): warning C4013: “is_a_ge_zero_and_a_lt_b”undefined；excepted outer return  int
..\..\src\col2im.c(78): error C2065: “output_w”: Undeclared definition
..\..\src\col2im.c(81): error C2065: “kernel_col”: Undeclared definition
..\..\src\col2im.c(82): error C2065: “output_col”: Undeclared definition
..\..\src\col2im.c(82): error C2065: “output_w”: Undeclared definition
..\..\src\col2im.c(82): error C2065: “output_col”: Undeclared definition
..\..\src\col2im.c(82): error C2065: “output_col”: Undeclared definition
  compare.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  connected_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  convolutional_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  conv_lstm_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  cost_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  cpu_gemm.c
  crnn_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  crop_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  compiling...
  darknet.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  dark_cuda.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  data.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  deconvolutional_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  demo.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  detection_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  detector.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  dice.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  dropout_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  gaussian_yolo_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  gemm.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  getopt.c
  gettimeofday.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  go.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  gru_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  im2col.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  image.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  list.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  local_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  compiling...
  lstm_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  matrix.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  maxpool_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  network.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  nightmare.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  normalization_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  option_list.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  parser.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  region_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  reorg_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  reorg_old_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  rnn.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  rnn_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  rnn_vid.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  route_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  sam_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  scale_channels_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  shortcut_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  softmax_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  super.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  compiling...
  swag.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  tag.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  tree.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  upsample_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  utils.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  voxel.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  writing.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  yolo.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory
  yolo_layer.c
O:\darknet-xp\include\darknet.h(39): fatal error C1083: could not open include file:“cuda_runtime.h”: No such file or directory

generation failed。

time cost 00:00:04.56
========== : success 0 ，fail 1 ，skip 0  ==========

```
I could success compile this on a windows 10 with cuda 10 machine.

**Comment by shartoo** [comment]
We want to deploy an object detection algorithm on an old machine with poor hardware,as you know other framework such as Tensorflow or Torch is too heavy and better hardware required. Your repo is an excellent project,i once deployed this on a raspberry pie 4+ with ubuntu 18.04 and it ran well.

**Comment by shartoo** [comment]
The project was generated by cmake as configuration below, with no gpu or cudnn enabled

![image](https://user-images.githubusercontent.com/5716282/68357576-c73e9400-0150-11ea-8f5f-10359f9a16bf.png)

**Comment by AlexeyAB** [comment]
Which of 3 ways do you use?
1. https://github.com/AlexeyAB/darknet#how-to-compile-on-windows-using-cmake-gui
2. https://github.com/AlexeyAB/darknet#how-to-compile-on-windows-using-vcpkg
3. https://github.com/AlexeyAB/darknet#how-to-compile-on-windows-legacy-way

**Comment by shartoo** [comment]
The first one with cmake-gui, i don't have a gpu on windows xp.

**Comment by AlexeyAB** [comment]
But cmake-gui doesn't create `yolo_cpp_dll_no_gpu.sln`

cmake-gui creates `Darknet.sln` and `dark.vcxproj` so open `darknet/Darknet.sln` in MSVS, compile it, and it will create `dark.dll` and `dark.lib` in the`darknet/Release` directory.

**Comment by shartoo** [comment]
Can you supply the cmake file for `yolo_cpp_dll_no_gpu.sln` ?

**Comment by AlexeyAB** [comment]
cmake creates Darknet.sln and dark.vcxproj. dark.vcxproj is the same as yolo_cpp_dll_no_gpu.sln

**Comment by shartoo** [comment]
I did successful compile the `cpp_dll_no_gpu` solution on windows xp by a vs2015 whose target platform is windows xp  on windows 10. Then i ran the darkent with my configuration with weights,data,image and so on,it could successful load weight and made one prediction bboxes and sequened in error `cannot read memory 00xx..6f` ,there should be three prediction bboxes. I print the `get_detection_boxes` result in source code of darknet before return to python which seems every prediction were correct. I'll keep checking if there sth wrong with data type transfering error.

![微信截图_20191114112938](https://user-images.githubusercontent.com/5716282/68824596-59571700-06d2-11ea-9ce6-cbb50dd80fbe.png)

**Comment by AlexeyAB** [comment]
Try to use the latest Darknet version.

**Comment by shartoo** [comment]
The problem still exists..

**Comment by shartoo** [comment]
I'll try to write some c/c++ code to wrap the raw detection result to json or some other simple data type ,there is no other solution to handle current memory error now.
